### PR TITLE
Allow configuring linux serial ports

### PR DESCRIPTION
**Pull request by Williangalvani**
[Screencast from 2022-12-22 15-16-05.webm](https://user-images.githubusercontent.com/4013804/209201851-b1909db6-5843-413f-a856-54b2c31a44b3.webm)

 - [ ] ~hotplugging is not working~ but this is not a blueos issue.
 - [x] fetching and updating backend data needs some love

### CODE REVIEW

**Comment by Williangalvani** [comment]
it works! SERIAL2 is a bluart-connected GPS, on usb
image

File: core/frontend/src/components/autopilot/AutopilotSerialConfiguration.vue [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+          matches += 1
+          if (matches > 1) {
+            return 'Duplicated entry!'
+          }
+        }
+      }
+      const regex1 = /(tcpclient|udpclient|tcpin|udpin):(?<ip1>\d*).(?<ip2>\d*).(?<ip3>\d*).(?<ip4>\d*):(?<port>\d*)$/
+      if (regex1.test(input)) {
+        const match = regex1.exec(input)
+        const ip1 = parseInt(match?.groups?.ip1 ?? '-1', 10)
```
Comment (patrickelectric) [inline core/frontend/src/components/autopilot/AutopilotSerialConfiguration.vue:161]: You can use `isIpAddress`

**Review by patrickelectric — CHANGES_REQUESTED** [review]
- It would be better if we could hide this configuration under a collapse like `Firmware update`:

image

- It should also be possible to configure the serial ports with SITL, no ?

File: core/frontend/src/components/autopilot/AutopilotSerialConfiguration.vue [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+          new_list.push({ port, endpoint })
+        }
+      }
+      return new_list
+    },
+  },
+  watch: {
+    current_serial_ports: {
+      handler(new_ports: SerialEndpoint[]) {
+        const ports = new_ports
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/frontend/src/components/autopilot/AutopilotSerialConfiguration.vue:118]: I probably console.logged it at a point and forget to take the intermediate out. fixing it.

File: core/services/ardupilot_manager/typedefs.py [hunk 3]
```diff
@@ -163,3 +166,38 @@ class EndpointType(str, Enum):
...
+            return value
+        raise ValueError(f"Invalid serial port: {value}. These must be between B and H. A is reserved.")
+
+    @validator("endpoint")
+    def valid_endpoint(cls: Any, value: str) -> str:
+        if Path(value).exists():
+            return value
+        if re.compile(r"tcp:\d*:wait$").match(value):
+            return value
+        matches = re.compile(r"(tcpclient|udpclient|tcpin|udpin):(?P<ip>(\d*\.){3}\d+):(?P<port>\d*)$").match(value)
```
Comment (patrickelectric) [inline core/services/ardupilot_manager/typedefs.py:195]: Should we also support `serial` ? `serial:/dev/ttyUSB2:115200`

**Review by patrickelectric — CHANGES_REQUESTED** [review]

File: core/services/ardupilot_manager/main.py [hunk 4]
```diff
@@ -96,6 +96,18 @@ def update_endpoints(endpoints: Set[Endpoint] = Body(...)) -> Any:
     autopilot.update_endpoints(endpoints)
 
 
+@app.put("/serials", status_code=status.HTTP_200_OK)
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/ardupilot_manager/main.py:99]: https://www.rfc-editor.org/rfc/rfc7231#section-4.3.4

image

I think put makes sense. but post would do too

File: core/services/ardupilot_manager/typedefs.py [hunk 5]
```diff
@@ -163,3 +166,38 @@ class EndpointType(str, Enum):
...
+    --serial4 /dev/ttyAMA2
+    --serial5 /dev/ttyAMA3
+    """
+
+    port: str
+    endpoint: str
+
+    @validator("port")
+    def valid_number(cls: Any, value: str) -> str:
+        if value in "BCDEFGH" and len(value) == 1:
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/ardupilot_manager/typedefs.py:185]: I think the original is easier to read.

File: core/services/ardupilot_manager/typedefs.py [hunk 6]
```diff
@@ -163,3 +166,38 @@ class EndpointType(str, Enum):
...
+    --serial3 /dev/ttyAMA1
+    --serial4 /dev/ttyAMA2
+    --serial5 /dev/ttyAMA3
+    """
+
+    port: str
+    endpoint: str
+
+    @validator("port")
+    def valid_number(cls: Any, value: str) -> str:
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/ardupilot_manager/typedefs.py:184]: yep!

File: core/services/ardupilot_manager/typedefs.py (hunk 3 above)
Comment (Williangalvani) [reply core/services/ardupilot_manager/typedefs.py:195]: is there an advantage in doing it? it sounds like additional work with not much to gain.

File: core/services/ardupilot_manager/ArduPilotManager.py [hunk 7]
```diff
@@ -131,10 +157,7 @@ def start_navigator(self, board: FlightController) -> None:
             f" -A udp:{master_endpoint.place}:{master_endpoint.argument}"
             f" --log-directory {self.settings.firmware_folder}/logs/"
             f" --storage-directory {self.settings.firmware_folder}/storage/"
-            f" -C /dev/ttyS0"
-            f" -B /dev/ttyAMA1"
-            f" -E /dev/ttyAMA2"
-            f" -F /dev/ttyAMA3"
+            f" {self.get_serial_cmdlines()}"
```
[the remark this answers is not in the document]
Comment (Williangalvani) [reply core/services/ardupilot_manager/ArduPilotManager.py:170]: makes sense, done

File: core/frontend/src/views/Autopilot.vue [hunk 8]
```diff
@@ -137,6 +149,14 @@ export default Vue.extend({
         'Vehicle type': this.vehicle_type ?? 'Unknown',
       }
     },
+    isLinuxFlightController(): boolean {
+      // this is setup this way so we can include other linux boards in the list in the future
+      const boardname = this.current_board?.name
+      if (!boardname) {
+        return false
+      }
+      return ['Navigator'].includes(boardname)
```
Comment (patrickelectric) [inline core/frontend/src/views/Autopilot.vue:158]: We should detect SITL.

File: core/services/ardupilot_manager/typedefs.py (hunk 3 above)
Comment (patrickelectric) [reply core/services/ardupilot_manager/typedefs.py:195]: Well, it's a valid endpoint, just saying

**Review by patrickelectric — COMMENTED** [review]
The data is available on the first API call but somehow is ignored, and the values only loads after more than 10 seconds.

https://user-images.githubusercontent.com/1215497/212366185-1bb2484b-5fe5-4cfa-b062-0bcf2317c440.mp4

**Comment by patrickelectric** [comment]
Docker says that the image is two days old, but the latest commit is from 16 hours ago.
image

**Comment by patrickelectric** [comment]
image

**Review by patrickelectric — APPROVED** [review]

**Pull request merged**